### ISSUE TITLE
keras Model with a dictionary as output: compile and fit expect layers names instead of the output dictionary's keys

### ISSUE THREAD

**Issue by durandg12** [tensorflow/tensorflow#34691]
- Have I written custom code (as opposed to using a stock example script provided in TensorFlow): yes
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04): macOS 10.13.6
- TensorFlow installed from (source or binary): pip install tensorflow==2.0.0-beta1
- TensorFlow version (use command below): v2.0.0-rc2-26-g64c3d382ca 2.0.0
- Python version: v3.6.7:6ec5cf24b7, Oct 20 2018, 03:02:14

**Describe the current behavior**

Using the functional Keras API, I build a model which has multiple outputs that are in a dictionary:
```
def make_model():
    inp = tf.keras.Input(1)
    y = tf.keras.layers.Dense(1)(inp)
    y_times_2 = tf.math.multiply(2., y)
    model = tf.keras.Model(inputs=inp, outputs={'y':y, 'y_times_2':y_times_2})
    return model
```
When compiling the model, I want to pass as the `loss` argument the following dictionary:
`{'y':'mean_squared_error', 'y_times_2':'mean_squared_error'}` 
and as the `metrics` argument the following dictionary:
`{'y':'mean_absolute_error', 'y_times_2':'mean_absolute_error'}`
as it is both intuitive and suggested [by the documentation of `compile`](https://www.tensorflow.org/versions/r2.0/api_docs/python/tf/keras/Model#compile) (see what is written for the `metrics` argument).
Similarly, for training the model, I want to pass as the `y` argument of `fit` the following dictionary:
`{'y':y, 'y_times_2':2*y}`.

It turns out that none of this works and using any of those dictionaries throws an error, because, strangely, the API does not expect `'y'` and `'y_times2'` as keys, whereas those are the keys of the output of the model, but it expects the name of the layers producing the outputs! For the `y` output, the layer producing it is the `Dense` layer defined in the model. For `y_times2` the layer producing it is an instance of `TensorFlowOpLayer` produced by the call of `tf.math.multiply` in the model definition. So, for example, the correct dictionary to use as the `loss` argument of `compile` is:
`{'dense':'mean_squared_error', 'tf_op_layer_Mul':'mean_squared_error'}`

Note that given that `TensorFlowOpLayer` layers are named automatically (even if I use the `name` argument of `tf.math.multiply`, because the name of the `TensorFlowOpLayer` layer is not the same as the name of the op), it can be really difficult to know the proper names to pass to `compile` and `fit` in a more complex example.

**Describe the expected behavior**

The expected behaviour should be as described in [the documentation of `compile`](https://www.tensorflow.org/versions/r2.0/api_docs/python/tf/keras/Model#compile) and should use the output names.

**Code to reproduce the issue**

```
import tensorflow as tf
import numpy as np

print('Using Tensorflow version {} (git version {})'.format(tf.version.VERSION, tf.version.GIT_VERSION))

x = np.random.normal(size=(64,))
y = x

def make_model():
    inp = tf.keras.Input(1)
    y = tf.keras.layers.Dense(1)(inp)
    y_times_2 = tf.math.multiply(2., y)
    model = tf.keras.Model(inputs=inp, outputs={'y':y, 'y_times_2':y_times_2})
    return model

model = make_model()

model.summary()

try:
    print('Trying compilation with the output names')
    model.compile(optimizer=tf.keras.optimizers.Adam(0.1), 
                  loss={'y':'mean_squared_error', 'y_times_2':'mean_squared_error'})
except Exception as e:
    print('The compilation failed with the following message error:')
    print(e)
    print('Trying compilation with the layers names')
    model.compile(optimizer=tf.keras.optimizers.Adam(0.1),
                  loss={'dense':'mean_squared_error', 'tf_op_layer_Mul':'mean_squared_error'})
    
try:
    print('\nTrying training with the output names')
    model.fit(x, {'y':y, 'y_times_2':2*y}, epochs=2)
except Exception as e:
    print('The training failed with the following message error:')
    print(e)
    print('Trying training with the layers names')
    model.fit(x, {'dense':y, 'tf_op_layer_Mul':2*y}, epochs=2)
    
print('\n###################\nAdd metrics to the example:')

model = make_model()

model.summary()

try:
    print('Trying compilation with the output names')
    model.compile(optimizer=tf.keras.optimizers.Adam(0.1), 
                  loss={'dense_1':'mean_squared_error', 'tf_op_layer_Mul_1':'mean_squared_error'},
                  metrics={'y':'mean_absolute_error', 'y_times_2':'mean_absolute_error'})
except Exception as e:
    print('The compilation failed with the following message error:')
    print(e)
    print('Trying compilation with the layers names')
    model.compile(optimizer=tf.keras.optimizers.Adam(0.1),
                  loss={'dense_1':'mean_squared_error', 'tf_op_layer_Mul_1':'mean_squared_error'},
                  metrics={'dense_1':'mean_absolute_error', 'tf_op_layer_Mul_1':'mean_absolute_error'})

model.fit(x, {'dense_1':y, 'tf_op_layer_Mul_1':2*y}, epochs=2)
```

with the following output:
```
Using Tensorflow version 2.0.0 (git version v2.0.0-rc2-26-g64c3d382ca)
Model: "model"
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
=================================================================
input_1 (InputLayer)         [(None, 1)]               0         
_________________________________________________________________
dense (Dense)                (None, 1)                 2         
_________________________________________________________________
tf_op_layer_Mul (TensorFlowO [(None, 1)]               0         
=================================================================
Total params: 2
Trainable params: 2
Non-trainable params: 0
_________________________________________________________________
Trying compilation with the output names
The compilation failed with the following message error:
Unknown entries in loss dictionary: ['y', 'y_times_2']. Only expected following keys: ['dense', 'tf_op_layer_Mul']
Trying compilation with the layers names

Trying training with the output names
The training failed with the following message error:
No data provided for "dense". Need data for each key in: ['dense', 'tf_op_layer_Mul']
Trying training with the layers names
Train on 64 samples
Epoch 1/2
64/64 [==============================] - 1s 9ms/sample - loss: 10.1650 - dense_loss: 2.0330 - tf_op_layer_Mul_loss: 8.1320
Epoch 2/2
64/64 [==============================] - 0s 62us/sample - loss: 7.1354 - dense_loss: 1.4271 - tf_op_layer_Mul_loss: 5.7084

###################
Add metrics to the example:
Model: "model_1"
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
=================================================================
input_2 (InputLayer)         [(None, 1)]               0         
_________________________________________________________________
dense_1 (Dense)              (None, 1)                 2         
_________________________________________________________________
tf_op_layer_Mul_1 (TensorFlo [(None, 1)]               0         
=================================================================
Total params: 2
Trainable params: 2
Non-trainable params: 0
_________________________________________________________________
Trying compilation with the output names
The compilation failed with the following message error:
Unknown entries in metrics dictionary: ['y', 'y_times_2']. Only expected following keys: ['dense_1', 'tf_op_layer_Mul_1']
Trying compilation with the layers names
Train on 64 samples
Epoch 1/2
64/64 [==============================] - 0s 5ms/sample - loss: 1.4019 - dense_1_loss: 0.2804 - tf_op_layer_Mul_1_loss: 1.1215 - dense_1_mean_absolute_error: 0.4108 - tf_op_layer_Mul_1_mean_absolute_error: 0.8217
Epoch 2/2
64/64 [==============================] - 0s 71us/sample - loss: 0.4672 - dense_1_loss: 0.0934 - tf_op_layer_Mul_1_loss: 0.3737 - dense_1_mean_absolute_error: 0.2408 - tf_op_layer_Mul_1_mean_absolute_error: 0.4816
```

**Comment by gadagashwini** [comment]
Could reproduce the issue with Tensorflow 2.0. 
Please see the [gist](https://colab.sandbox.google.com/gist/gadagashwini/7606fa53332f78d64286bae0e90d7b2a/untitled284.ipynb). Thanks!

**Comment by ghwatson** [comment]
Given the documentation on keras, I don't think you're supposed to pass a dictionary for outputs. I've never seen keras code in the tensorflow docs that do that.

**Comment by durandg12** [comment]
As I said, [the documentation of `compile`](https://www.tensorflow.org/versions/r2.0/api_docs/python/tf/keras/Model#compile) implies it, read what is written in  the `metrics` argument.

**Comment by ghwatson** [comment]
You're correct about what's written in the metrics argument. However, that's for the metrics argument not the outputs argument. I could be wrong as I'm new to Keras myself, but again, I haven't seen anything that specifically does that for the outputs argument. I've only seen it for the losses and metric arguments. I believe that it attaches these two things to inputs and outputs by looking for the names of the tensors they're supposed to be attached to via the key you provide. I don't think that it directly matches keys. Again, take with grain of salt though as I'm new myself!

**Comment by ghwatson** [comment]
Also, if you're convinced, I suspect there's some bugs in the dictionary-style API as I've described it. See my issue I submitted here: #34713

**Comment by durandg12** [comment]
I see what you mean, and it seems that you are right. The doc you linked in your issue supports that.

> I believe that it attaches these two things to outputs by looking for the names of the tensors they're supposed to be attached to via the key you provide.

From my experiments (and the doc you linked in your issue), it looks for the name of the layer outputting the tensors, not the name of the tensors. And this is an information that can be difficult to find. Furthermore, like shown in my example above, there are layers for which you don't have total control over the name (I think this corner case was not taken in consideration when designing the API).

 If, as you said, Keras should not be given a dict as output, then my issue becomes a feature request. However it works when I simply `call` the model on some data. With my example code above, if you then execute `model(tf.expand_dims(x, axis=-1))` you get the desired output: a `dict` of `Tensor` objects with keys `'y'`and `'y_times_2'`. Being able to name the outputs and having those names don't match the names you give to the `loss` and `metrics` arguments of `compile` seems pretty counterintuitive. I think that, if the ouputs are named by using a dictionary as output, those names should override the name of the layers producing the outputs.

**Comment by goldiegadde** [comment]
@durandg12 This issue is resolved with the latest tf-nightly and 2.2.0-rc2. here is a [colab gist](https://colab.research.google.com/gist/goldiegadde/0ad3fed10b0cbfb3633ee3ff05dcad31/github-issue-34691.ipynb)
Could you please verify and close this issue accordingly.

**Comment by durandg12** [comment]
@goldiegadde I would not say the issue is totally solved, looking at the gist. Althought my main points are solved, the information printed by the keras API during training still uses the name of the tensor producing the output instead of the name of the output given by the user:
```
Epoch 1/2
2/2 [==============================] - 0s 2ms/step - loss: 9.2561 - dense_loss: 1.8512 - tf_op_layer_Mul_loss: 7.4048
Epoch 2/2
2/2 [==============================] - 0s 1ms/step - loss: 6.7087 - dense_loss: 1.3417 - tf_op_layer_Mul_loss: 5.3670
```

**Comment by jvishnuvardhan** [comment]
@durandg12 I am not sure whether it is a workaround or not. I changed two lines in your code as shown below. With that change, now I can see the following output with the names given to the layers. Please take a look at the [gist here](https://colab.research.google.com/gist/jvishnuvardhan/7711f849ec186ce53fa8f8489845a7ea/github-issue-34691.ipynb).

```
    y = tf.keras.layers.Dense(1,name='y')(inp)
    y_times_2 = tf.keras.layers.Lambda(lambda y:tf.math.multiply(2., y),name='y_times_2')(y)
```

Output

```
Epoch 1/2
2/2 [==============================] - 0s 4ms/step - loss: 18.1492 - y_loss: 3.6298 - y_times_2_loss: 14.5194 - y_mean_absolute_error: 1.6186 - y_times_2_mean_absolute_error: 3.2371
Epoch 2/2
2/2 [==============================] - 0s 6ms/step - loss: 13.7314 - y_loss: 2.7463 - y_times_2_loss: 10.9851 - y_mean_absolute_error: 1.4094 - y_times_2_mean_absolute_error: 2.8189
<tensorflow.python.keras.callbacks.History at 0x7f6b66c5f390>
```
